swarm/src/handler.rs

```diff
@@ -328,118 +323,9 @@ pub struct AddressChange<'a> {
-            protocols: actually_added_protocols,
-        }))
-    }
-
-    /// Compute the [`ProtocolsChange`] that results from removing `to_remove` from `existing_protocols`.
-    ///
-    /// Returns `None` if the change is a no-op, i.e. none of the protocols in `to_remove` are in `existing_protocols`.
-    pub(crate) fn remove(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        to_remove: &'a HashSet<StreamProtocol>,
-    ) -> Option<Self> {
-        let mut actually_removed_protocols = existing_protocols.intersection(to_remove).peekable();
-
-        actually_removed_protocols.peek()?;
-
-        Some(ProtocolsChange::Removed(ProtocolsRemoved {
-            protocols: Either::Right(actually_removed_protocols),
-        }))
-    }
-
-    /// Compute the [`ProtocolsChange`]s required to go from `existing_protocols` to `new_protocols`.
-    pub(crate) fn from_full_sets(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        new_protocols: &'a HashSet<StreamProtocol>,
-    ) -> SmallVec<[Self; 2]> {
-        if existing_protocols == new_protocols {
-            return SmallVec::new();
-        }
-
-        let mut changes = SmallVec::new();
-
-        let mut added_protocols = new_protocols.difference(existing_protocols).peekable();
-        let mut removed_protocols = existing_protocols.difference(new_protocols).peekable();
-
-        if added_protocols.peek().is_some() {
-            changes.push(ProtocolsChange::Added(ProtocolsAdded {
-                protocols: added_protocols,
-            }));
-        }
-
-        if removed_protocols.peek().is_some() {
-            changes.push(ProtocolsChange::Removed(ProtocolsRemoved {
-                protocols: Either::Left(removed_protocols),
-            }));
-        }
-
-        changes
-    }
-}
-
-/// An [`Iterator`] over all protocols that have been added.
-#[derive(Debug, Clone)]
-pub struct ProtocolsAdded<'a> {
-    protocols: Peekable<Difference<'a, StreamProtocol, RandomState>>,
-}
-
-impl<'a> ProtocolsAdded<'a> {
-    pub(crate) fn from_set(protocols: &'a HashSet<StreamProtocol, RandomState>) -> Self {
-        ProtocolsAdded {
-            protocols: protocols.difference(&EMPTY_HASHSET).peekable(),
-        }
-    }
-}
-
-/// An [`Iterator`] over all protocols that have been removed.
-#[derive(Debug, Clone)]
-pub struct ProtocolsRemoved<'a> {
-    protocols: Either<
-        Peekable<Difference<'a, StreamProtocol, RandomState>>,
-        Peekable<Intersection<'a, StreamProtocol, RandomState>>,
-    >,
-}
-
-impl<'a> ProtocolsRemoved<'a> {
-    #[cfg(test)]
-    pub(crate) fn from_set(protocols: &'a HashSet<StreamProtocol, RandomState>) -> Self {
-        ProtocolsRemoved {
-            protocols: Either::Left(protocols.difference(&EMPTY_HASHSET).peekable()),
-        }
-    }
-}
-
-impl<'a> Iterator for ProtocolsAdded<'a> {
-    type Item = &'a StreamProtocol;
-    fn next(&mut self) -> Option<Self::Item> {
-        self.protocols.next()
-    }
-}
```

Comment:
It was by design that this is the only public API exposed by `ProtocolsAdded` and `ProtocolsRemvoed`. Can't we keep this but back it by a `SmallVec` instead?